source/common/filesystem/kqueue/watcher_impl.cc

```diff
@@ -0,0 +1,170 @@
+  FileWatchPtr watch = addWatch_(path, events, cb, false);
+  if (watch == nullptr) {
+    throw EnvoyException(fmt::format("invalid watch path {}", path));
+  }
+}
+
+WatcherImpl::FileWatchPtr WatcherImpl::addWatch_(const std::string& path, uint32_t events,
+                                                 Watcher::OnChangedCb cb, bool path_must_exist) {
+  bool watching_dir = false;
+  int watch_fd = open(path.c_str(), O_SYMLINK);
+  if (watch_fd == -1) {
+    if (path_must_exist) {
+      return nullptr;
+    }
+
+    size_t last_slash = path.rfind('/');
+    if (last_slash == std::string::npos) {
+      throw EnvoyException(fmt::format("invalid watch path {}", path));
+    }
+
+    std::string directory = path.substr(0, last_slash);
+    watch_fd = open(directory.c_str(), 0);
+    if (watch_fd == -1) {
+      return nullptr;
+    }
+
+    watching_dir = true;
+  }
+
+  FileWatchPtr watch(new FileWatch());
+  watch->fd_ = watch_fd;
+  watch->file_ = path;
+  watch->events_ = events;
+  watch->callback_ = cb;
+  watch->watching_dir_ = watching_dir;
+
+  int flags = NOTE_DELETE | NOTE_RENAME;
+  if (watching_dir) {
+    flags = NOTE_DELETE | NOTE_WRITE;
+  }
+
+  struct kevent event;
+  EV_SET(&event, watch_fd, EVFILT_VNODE, EV_ADD | EV_CLEAR, flags, 0,
+         reinterpret_cast<void*>(watch_fd));
+
+  if (kevent(queue_, &event, 1, NULL, 0, NULL) == -1) {
+    throw EnvoyException(
+        fmt::format("unable to add filesystem watch for file {}: {}", path, strerror(errno)));
+  }
+
+  if (event.flags & EV_ERROR) {
+    throw EnvoyException(
+        fmt::format("unable to add filesystem watch for file {}: {}", path, strerror(event.data)));
+  }
+
+  ENVOY_LOG(debug, "added watch for file: '{}' fd: {}", path, watch_fd);
+
+  watches_[watch_fd] = watch;
+
+  return watch;
+}
+
+void WatcherImpl::removeWatch(FileWatchPtr& watch) {
+  int fd = watch->fd_;
+  watches_.erase(fd);
+}
+
+void WatcherImpl::onKqueueEvent() {
+  struct kevent event = {};
+  timespec nullts = {0, 0};
+
+  while (true) {
+    uint32_t events = 0;
+    int nevents = kevent(queue_, NULL, 0, &event, 1, &nullts);
+    if (nevents < 1 || event.udata == nullptr) {
+      return;
+    }
+
+    int watch_fd = reinterpret_cast<std::intptr_t>(event.udata);
+
+    FileWatchPtr file = watches_[watch_fd];
+    if (file == nullptr) {
+      continue;
+    }
+    ASSERT(watch_fd == file->fd_);
+
+    if (file->watching_dir_) {
+      if (event.fflags & NOTE_DELETE) {
+        // directory was deleted
+        removeWatch(file);
+        return;
+      }
+
+      if (event.fflags & NOTE_WRITE) {
+        // directory was written -- check if the file we're actually watching appeared
+        FileWatchPtr new_file = addWatch_(file->file_, file->events_, file->callback_, true);
+        if (new_file != nullptr) {
+          removeWatch(file);
+          file = new_file;
+
+          events |= Events::MovedTo;
```

Comment:
`events` is only ever 0 or Events::MovedTo (which is the only value defined in `include/envoy/filesystem/filesystem.h`). The inotify implementation does this as well. I assumed it was because someone expected to add more event types some day.